py/server/deephaven/table_reader.py

```diff
@@ -0,0 +1,96 @@
+    names to numpy arrays.
+
+    Notes,
+        1) If the table is refreshing and no update graph locks are currently being held, in order to provide a
+            consistent view of the data, the table reader will try to acquire the shared lock of the update graph
+            before start reading the table data. The side effect of this is that the table will not be able to refresh
+            while the table reader is being used. The acquired shared lock will be released after the table reader is
+            destroyed either implicitly when used in a for-loop statement, or explicitly by setting it to None or using
+            the del statement or calling the close() method on it.
+
+        2) The returned generator maintains a context internally and only guarantees the release of the context when
+            it is used in a for-loop statement. If the generator is used outside a for-loop statement by calling next()
+            on it, the context will not automatically be released. To avoid resource leaks, the generator should be
+            closed after use either by setting it to None or using the del statement or calling the close() method on
+            it.
+
+    Args:
+        table (Table): The table to read.
+        cols (Optional[Union[str, Sequence[str]]]): The columns to read. If None, all columns are read.
+        chunk_size (Optional[int]): The number of rows to read at a time. Default is None, meaning all rows in the table
+            are read in one chunk.
+
+    Returns:
+        A generator that yields a dictionary of column names to numpy arrays.
+    """
+    return _table_reader_by_row_set(table, cols, table.j_table.getRowSet(), chunk_size, False)
+
+
+def _table_reader_by_row_set(table: Table, cols: Optional[Union[str, Sequence[str]]], row_set: jpy.JType, chunk_size: Optional[int],
+                             prev: bool = False) -> Generator[Dict[str, np.ndarray], None, None]:
+    """ A generator that reads the chunks of rows from a table into a dictionary.  The dictionary is a map of column
+    names to numpy arrays.
+
+    Args:
+        table (Table):  The table to read.
+        cols (Optional[Union[str, Sequence[str]]]): The columns to read. If None, all columns are read.
+        row_set (jpy.JType): The row set to read.
+        chunk_size (Optional[int]): The number of rows to read at a time. If None, all rows in the row set are read.
+        prev (bool): If True, read the previous values. Default is False.
+
+    Returns:
+        A generator that yields a dictionary of column names to numpy arrays.
+    """
+    col_defs = _col_defs(table, cols)
+
+    row_sequence_iterator = row_set.getRowSequenceIterator()
+    col_sources = [table.j_table.getColumnSource(col_def.name) for col_def in col_defs]
+    chunk_size = row_set.size() if not chunk_size else chunk_size
+    j_reader_context = _JTableUpdateDataReader.makeContext(chunk_size, *col_sources)
+    with update_graph.auto_locking_ctx(table):
```

Comment:
So this holds until we yield for the last time? And if you delete the generator, that actually causes the finally to execute? Seems surprising.

Reply:
I wanted to test this, so I have some example code at https://github.com/deephaven/deephaven-core/pull/5595#discussion_r1643276710